### PR TITLE
gh-91616: re module, fix .fullmatch() mismatch when using Atomic Grouping or Possessive Quantifiers

### LINKED ISSUE #91616: Path.rglob fails with *.* on Python 3.11.0a7
**Bug report**

On Python 3.11.0a7, calling `Path.rglob("*.*")`  returns an iterator over no files, whereas on previous versions (including alpha 6) it would correctly return an iterator over all files in the directory and all of its children.

The following code illustrates this problem:

```python
import pathlib, tempfile, pprint

with tempfile.TemporaryDirectory() as tmpdir:
	tmp_path = pathlib.Path(tmpdir)
	(tmp_path / "a.txt").touch()
	(tmp_path / "b.txt").touch()
	(tmp_path / "c").mkdir()
	(tmp_path / "c" / "d.txt").touch()
	pprint.pprint(list(tmp_path.rglob("*.*")))
	pprint.pprint(list(tmp_path.rglob("*.txt")))
```

Save this as `pathlib_rglob.py`.

**Output:**

```console

$ python3.11 pathlib_rglob.py 
[]
[PosixPath('/tmp/tmpngjo2cyn/a.txt'),
 PosixPath('/tmp/tmpngjo2cyn/b.txt'),
 PosixPath('/tmp/tmpngjo2cyn/c/d.txt')]
$ python3.10 pathlib_rglob.py 
[PosixPath('/tmp/tmpodfbqyhx/a.txt'),
 PosixPath('/tmp/tmpodfbqyhx/b.txt'),
 PosixPath('/tmp/tmpodfbqyhx/c/d.txt')]
[PosixPath('/tmp/tmpodfbqyhx/a.txt'),
 PosixPath('/tmp/tmpodfbqyhx/b.txt'),
 PosixPath('/tmp/tmpodfbqyhx/c/d.txt')]

$ # This is Python 3.11.0a6 built from source
$ ~/python311a6/bin/python3.11 pathlib_rglob.py 
[PosixPath('/tmp/tmpobellcrw/a.txt'),
 PosixPath('/tmp/tmpobellcrw/b.txt'),
 PosixPath('/tmp/tmpobellcrw/c/d.txt')]
[PosixPath('/tmp/tmpobellcrw/a.txt'),
 PosixPath('/tmp/tmpobellcrw/b.txt'),
 PosixPath('/tmp/tmpobellcrw/c/d.txt')]
```

-----

After much debugging I have tracked the issue to https://github.com/python/cpython/pull/32029
With that change, `fnmatch.translate("*.*")` gives `(?s:(?>.*?\.).*)\Z`.
With the change reverted (or on 3.11.0a6) its `(?s:(?=(?P<g0>.*?\.))(?P=g0).*)\Z`.
I don't know what the difference between those amounts to, but the former doesn't work while the latter does:

```python3
>>> import re
>>> re.fullmatch('(?s:(?>.*?\\.).*)\\Z', "a.txt")
>>> re.fullmatch('(?s:(?=(?P<g0>.*?\\.))(?P=g0).*)\\Z', "a.txt")
<re.Match object; span=(0, 5), match='a.txt'>
```

**Your environment**

- CPython versions tested on: 3.10.4, 3.11.0a6, 3.11.0a7, built from git checkouts of the relevant tag.
- Operating system and architecture: Ubuntu 20.04 amd64

### PR DESCRIPTION
**Pull request by ghost**
These jumps should use `DO_JUMP0()` instead of `DO_JUMP()`:
- `JUMP_POSS_REPEAT_1`
- `JUMP_POSS_REPEAT_2`
- `JUMP_ATOMIC_GROUP`

Closes #91616.

### CODE REVIEW

File: Lib/test/test_pathlib.py [hunk 1]
```diff
@@ -1412,6 +1413,8 @@ def cleanup():
             f.write(b"this is file B\n")
         with open(join('dirC', 'fileC'), 'wb') as f:
             f.write(b"this is file C\n")
+        with open(join('dirC', 'novel.txt'), 'w', encoding='ascii') as f:
+            f.write("this is a novel\n")
```
Comment (serhiy-storchaka) [inline Lib/test/test_pathlib.py]: You can just open the file in binary mode.

File: Lib/test/test_pathlib.py [hunk 2]
```diff
@@ -1689,7 +1695,7 @@ def test_rglob_symlink_loop(self):
         expect = {'brokenLink',
                   'dirA', 'dirA/linkC',
                   'dirB', 'dirB/fileB', 'dirB/linkD',
-                  'dirC', 'dirC/dirD', 'dirC/dirD/fileD', 'dirC/fileC',
+                  'dirC', 'dirC/dirD', 'dirC/dirD/fileD', 'dirC/fileC', 'dirC/novel.txt',
```
Comment (serhiy-storchaka) [inline Lib/test/test_pathlib.py]: Too long line.

File: Lib/test/test_re.py [hunk 3]
```diff
@@ -2238,6 +2254,34 @@ def test_findall_atomic_grouping(self):
         self.assertEqual(re.findall(r'(?>(?:ab)?)', 'ababc'), ['ab', 'ab', '', ''])
         self.assertEqual(re.findall(r'(?>(?:ab){1,3})', 'ababc'), ['abab'])
 
+    def test_bug_gh91616(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_re.py:2257]: Aren't these test redundant? New assertions in `test_fullmatch_possessive_quantifiers` and `test_fullmatch_atomic_grouping` cover all changes.

File: Lib/test/test_pathlib.py [hunk 4]
```diff
@@ -1679,6 +1682,9 @@ def _check(glob, expected):
         p = P(BASE, "dirC")
         _check(p.rglob("file*"), ["dirC/fileC", "dirC/dirD/fileD"])
         _check(p.rglob("*/*"), ["dirC/dirD/fileD"])
+        # gh-91616, a re module regression
+        _check(p.rglob("*.txt"), ["dirC/novel.txt"])
+        _check(p.rglob("*.*"), ["dirC/novel.txt"])
```
Comment (serhiy-storchaka) [inline Lib/test/test_pathlib.py:1687]: Would not testing `"*eC"` be enough?

File: Lib/test/test_pathlib.py (hunk 1 above)
Comment (ghost) [reply Lib/test/test_pathlib.py]: done

File: Lib/test/test_pathlib.py (hunk 2 above)
Comment (ghost) [reply Lib/test/test_pathlib.py]: done

File: Lib/test/test_pathlib.py (hunk 4 above)
Comment (ghost) [reply Lib/test/test_pathlib.py:1687]: I would like to keep this.
`*.txt` is a common usage, but it's not tested for `.rglob()`.

File: Lib/test/test_re.py (hunk 3 above)
Comment (ghost) [reply Lib/test/test_re.py:2257]: done

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**